benchmarks/distributed/ddp/compare_ddp.py

```diff
@@ -0,0 +1,150 @@
+import numpy as np
+import os
+import python_ddp
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+import torch.nn as nn
+import torch.optim as optim
+import torchvision.models as models
+
+from enum import Enum
+from torch.nn.parallel import DistributedDataParallel as DDP
+
+class DDPOption(Enum):
+    DDP_CPP_CORE = 1
+    LEGACY_DISTRIBUTED_DATA_PARALLEL = 2
+    PYTHON_DDP_SYNC = 3
+    PYTHON_DDP_ASYNC = 4
+
+def setup(rank, world_size):
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = '12355'
+
+    # initialize the process group
+    dist.init_process_group("gloo", rank=rank, world_size=world_size)
+
+def create_ddp_model(module, rank, pg, ddp_option, buffer_size):
+    """Helper to create DDPModel. """
+    if ddp_option == DDPOption.DDP_CPP_CORE:
+        ddp_model = DDP(module, device_ids=[rank], process_group=pg)
+        ddp_model._set_static_graph()
+        return ddp_model
+    elif ddp_option == DDPOption.LEGACY_DISTRIBUTED_DATA_PARALLEL:
+        return legacy_ddp.LegacyDistributedDataParallel(module, pg, buffer_size)
+    elif ddp_option == DDPOption.PYTHON_DDP_SYNC:
+        return python_ddp.PythonDDP(module, pg, False, buffer_size)
+    elif ddp_option == DDPOption.PYTHON_DDP_ASYNC:
+        return python_ddp.PythonDDP(module, pg, True, buffer_size)
+    else:
+        raise NotImplementedError
+
+def run_ddp(rank, world_size, epochs, ddp_option, buffer_size):
+    print(f'Invoked run_ddp rank {rank}')
+
+    # Setup
+    print("setting up ... ")
+    setup(rank, world_size)
+    torch.manual_seed(rank)
+    torch.cuda.manual_seed(rank)
+    device = torch.device('cuda:%d' % rank)
+    print('done')
+
+    # Create ResNet50 module and wrap in DDP module.
+    pg = dist.distributed_c10d._get_default_group()
+    model = models.resnet50().to(device)
+    ddp_model = create_ddp_model(model, rank, pg, ddp_option, buffer_size)
+    assert ddp_model is not None
+
+    loss_fn = nn.MSELoss()
+    optimizer = optim.SGD(ddp_model.parameters(), lr=0.001)
+
+    # Container to hold: event -> list of events in milliseconds
+    MODEL_FORWARD = "forward"
+    MODEL_BACKWARD = "backward"
+    metrics = {MODEL_FORWARD: [], MODEL_BACKWARD: []}
+
+    for epoch in range(epochs):
+        if epoch % 10 == 0:
+            print(f'Epoch {epoch}/{epochs} ...')
+
+        start = torch.cuda.Event(enable_timing=True)
+        end = torch.cuda.Event(enable_timing=True)
+
+        # TODO(bowangbj): Switch to real training set from ImageNet.
+        inputs = torch.rand([32, 3, 224, 224], device=device)
+        labels = torch.rand([32, 1000], device=device)
+
+        # Forward
+        start.record()
+        outputs = ddp_model(inputs)
+        end.record()
+        torch.cuda.synchronize()
+        metrics[MODEL_FORWARD].append(start.elapsed_time(end))
+
+        # Backward
+        start.record()
+        loss_fn(outputs, labels).backward()
```

Comment:
nit: `loss_fn()` should be part of forward.

Reply:
Done.  maybe in distributed world, backward is more like "reduction" across processes ?

Reply:
After a second thought, it might still make sense to have loss_func.backward in the backward step. 

Otherwise, it seems there will be nothing to measure in the backward step for production DDP?  Just found that the delay for DDP backward is < 1ms since it has nothing to do if loss func is not included :)

Reply:
oh, this is surprising and doesn't look right. We previously measured things like below, and usually backward is 3x-5x slower than the forward, and it will be even slower when there are communications. See Fig. 6 in http://www.vldb.org/pvldb/vol13/p3005-li.pdf

```python
fwd_start.record()
loss = loss_fn(model(inputs), labels)
fwd_stop.record()

bwd_start.record()
loss.backward()
bwd_stop.record()
```

If you consistently see this behavior (backward delay is negligible compared to forward), could you please share a small repro?

Reply:
Sry I misread your cmt. 

I misplaced 
loss_fn(outputs, labels).backward()   -- in forward step

forward: 
outputs = model(inputs)
loss_fn(outputs, labels).backward()

backward:
nothing    :)

Changed it back to sth like 
forward:
outputs = models(inputs)
loss = loss_fn(labels, outputs)

backward:
loss.backward()

Tried it a few times, moving loss_fn(labels, outputs) would not make a big difference as expected.